src/framework/operations.hpp

```diff
@@ -683,6 +723,20 @@ json_t op_to_json(const Op &op) {
 }
 
 
+void to_json(json_t &js, const OpType& type) {
+  std::stringstream ss;
+  ss << type;
+  js = ss.str();
+}
+
+
+void to_json(json_t &js, const DataSubType& subtype) {
+  std::stringstream ss;
+  ss << subtype;
+  js = ss.str();
+}
```

Comment:
Actually, come to think of it, it could just be that `DataType` and `DataSubType` should be one single struct with three fields:
```c++
enum {
    Stabilizer,
    Clifford,
    Statevector,
    ...,
} DataSaveType;

enum {
    List,
    Accumulate,
    Average,
    ...,
} DataReductionType;

struct {
    DataSaveType type,
    DataReductionType reduction,
    bool conditional,
} DataType;
```

or something like that?

Reply:
This makes switch statements where they are used more difficult since you can't do 1 switch statement, but have to nest them.

Basically this code is just to convert the enum class value to a string for metadata. Which is surprisingly difficult to do in C++ hence just using JSON library to do it instead.

Reply:
I'm not sure I understand why structuring the data like this makes things more difficult?  You've already got the split between `DataType` and `DataSubType`, so this doesn't add any more switches.  In each switch you already need to separate cases `list` and `c_list` (etc), so it's no more complex between having the inner bit be a `?:` or similar instead, and structuring the data clearly shows the association between `list` and `c_list` in the data.  If anything this _reduces_ the number of lines of code, because, for example:
```c++
switch (type) {
    case DataSubType::list:
      result.data.add_list(datum, key);
      break;
    case DataSubType::c_list:
      result.data.add_list(datum, key, creg_.memory_hex());
      break;
 }
```
would become
```c++
switch (type.reduction) {
    case DataSubType::list:
      type.conditional
        ? result.data.add_list(datum, key, creg_.memory_hex())
        : result.data.add_list(datum, key);
      break;
 }
```
and the structured data has lots of other benefits, like ensuring that everything's passed as one and can't get out-of-sync.  If it's not reasonable to join `DataSaveType` and `DataReductionType` I understand, but I'm not a fan of mixing a conditional flag into a related enumeration when it's just as easy to structure the data fully.

Reply:
I will put this suggestion in the "it would be nice, but i don't want to do this large a refactor right now" category.